contributors/design-proposals/storage/csi-migration.md

```diff
@@ -0,0 +1,443 @@
+aware/dependent on what kind of PV ends up getting bound to their PVC (this is a
+core tenet of the k8s storage system).
+
+If the volume was provisioned by the in-tree provisioner but CSI Migration is
+now turned on - On Delete, the internal volume plugin will translate in-tree
+volume source into CSI Source and put it as an annotation to the PV if
+necessary. The external provisioner will check for
+“pv.kubernetes.io/provisioned-by” annotation as usual (added during provisioning
+either by internal or external provisioner) and presence of JSON with CSI
+Source. It will delete the PV only when the PV itself is CSI PV or it has JSON
+annotation with CSI source.
+
+
+### Non-Dynamic Provisioned Volumes (and volumes provisioned before migration)
+In the OperationGenerator at the start of each volume operation call we will
+check to see whether the plugin has been migrated. If it is migrated, we will
+call the in-tree plugin “translation” interface method to get the CSI Source
+JSON.
+
+For Controller calls, we will put the CSI Source JSON into an annotation on the
+PV PV.Annotations[‘k8s.io/CSIVolumeSource‘] and the VolumeToAttach object. This
+requires giving A/D controller and kubelet volume manager PV Update/Patch
+permissions. This is necessary because the CSI external controller components
+all GET the PV object and use it as a source of truth. We also foresee that
+future controller calls such as resize and snapshot will also use and update the
+PV as a source of truth. All these external components will also need to be
+updated to look at the PV Annotation to determine whether it is meant for that
+CSI Driver or not, and not just at the PV Source. 
+
+For Node calls, we can put the CSI Source JSON into an annotation on the
+in-memory PV in the VolumeToMount object. This is enough since the information
+in this object is propagated all the way through to the CSI components when
+SetUp is called.
+
+FindPluginBySpec will also have to be modified to return the CSI plugin if the
+plugin migration flag is set for this particular plugin and the annotation has
+been added.
+
+Then the rest of the code in the Operation Generator can execute as normal with
+the CSI Plugin and the annotation in the requisite locations.
+
+GetCSISourceFromSpec is used by both Node/Controller in-tree CSI functions and
+can be changed to get the CSI Source from the annotation
+PV.Annotations[‘k8s.io/CSIVolumeSource‘] first if it exists. This will ensure
+all the in-tree CSI plugin functions will be using the CSI Source translated
+from before.
+
+Caveat: For ALL detach calls of plugins that MAY have already been migrated we
+have to attempt to DELETE the VolumeAttachment object that would have been
+created if that plugin was migrated. This is because Attach after migration
+creates a VolumeAttachment object, and if for some reason we are doing a detach
+with the in-tree plugin, the VolumeAttachment object becomes orphaned.
```

Comment:
This would happen if user disables the CSI migration feature gate right?  Another thing to keep in mind is - when feature gate is disabled, both controller-manager and kubelet should be able to "recover" state without relying on CSI only objects. For example, `node.Status.VolumeAttached` property set for CSI plugin should be compatible with in-tree plugin , in case user downgrades to in-tree plugin.

Reply:
That is correct. With this proposal the `node.Status.VolumeAttached` property should be the same when migrated vs. not migrated so the state should be recoverable here.